File: 03-FZHTabbarController/03-FZHTabbarController/03-FZHTabbarController/FZHTabBar.m
//
//  FZHTabBar.m
//  03-FZHTabbarController
//
//  Created by 冯志浩 on 16/8/16.
//  Copyright © 2016年 FZH. All rights reserved.
//

#import "FZHTabBar.h"
#import "FZHTabbarButton.h"
@implementation FZHTabBar
-(NSMutableArray *)tabBarButtons
{
    if (_tabBarButtons== nil) {
        _tabBarButtons=[NSMutableArray array];
    }
    return _tabBarButtons;
}
-(id)initWithFrame:(CGRect)frame
{
    self=[super initWithFrame:frame];
    if (self) {
       
    }
    return self;
}
-(void)plusButtonClick
{
    if ([self.delegate respondsToSelector:@selector(tabBardidPlusButton:)]) {
        [self.delegate tabBardidPlusButton:self];
    }
}
-(void)addTabBarButtonWithItem:(UITabBarItem *)item
{
    //1.创建按钮
    FZHTabbarButton * button=[[FZHTabbarButton alloc]init];
    [self.tabBarButtons addObject:button];
    //2.设置数据
    button.item=item;
    
    //3.添加按钮
    [button addTarget:self action:@selector(buttonClick:) forControlEvents:UIControlEventTouchDown];
    [self addSubview:button];
    //4.默认选中
    if (self.tabBarButtons.count==1) {
        
        [self buttonClick:button];
    }
    
}
-(void)buttonClick:(FZHTabbarButton *)button
{
    //1.通知代理
    if ([self.delegate respondsToSelector:@selector(tabBar:didSelectedButtonFrom:to:)]) {
        [self.delegate tabBar:self didSelectedButtonFrom:(int)self.selectedButton.tag to:(int)button.tag];
    }
    //2.控制器选中按钮
    self.selectedButton.selected=NO;
    button.selected=YES;
    self.selectedButton=button;
}
//布局子控件
-(void)layoutSubviews
{
    [super layoutSubviews];
    
    //1. 4个按钮
    CGFloat buttonW=self.frame.size.width/self.subviews.count;
    CGFloat buttonH=self.frame.size.height;
    CGFloat buttonY = 0;
    
    for (int index=0; index<self.tabBarButtons.count; index++) {
        
        //1.取出按钮
        FZHTabbarButton * button=self.tabBarButtons[index];
        //2.设置按钮的frame
        CGFloat buttonX=index * buttonW;
        button.frame=CGRectMake(buttonX, buttonY, buttonW, buttonH);
        //3.绑定tag
        button.tag=index;
    }
}

@end

File: 03-FZHTabbarController/03-FZHTabBarController(swift)/03-FZHTabBarController(swift)/FZHTabbarController/View/FZHTabBar.swift
//
//  FZHTabBar.swift
//  03-FZHTabBarController(swift)
//
//  Created by 冯志浩 on 16/8/15.
//  Copyright © 2016年 FZH. All rights reserved.
//

import UIKit
protocol FZHTabBarDelegate {
    func tabbar(_ tabbar: FZHTabBar,formWhichItem: Int, toWhichItem: Int)
}
class FZHTabBar: UIView {
    var tabBarButtons: NSMutableArray = []
    var selectedButton = FZHTabbarButton()
    
    var fzhTabbarDelegate: FZHTabBarDelegate! = nil
    override init(frame: CGRect) {
        super.init(frame: frame)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func addTabbarButtonWith(_ item: UITabBarItem) -> Void {
        let button = FZHTabbarButton()
        button.item = item
        self.tabBarButtons.add(button)
        button.addTarget(self, action: #selector(buttonDidTouch), for: .touchDown)
       self.addSubview(button)
        //        默认选中
        if self.tabBarButtons.count == 1 {
            self.buttonDidTouch((self.tabBarButtons[0] as? FZHTabbarButton)!)
        }
    }
    
    func buttonDidTouch(_ button: FZHTabbarButton) -> Void {
        fzhTabbarDelegate.tabbar(self, formWhichItem: self.selectedButton.tag, toWhichItem: button.tag)
//        控制器选中按钮
        self.selectedButton.isSelected = false
        button.isSelected = true
        self.selectedButton = button
    }

    override func layoutSubviews() {
        super.layoutSubviews()
//        button
        let buttonW = self.frame.size.width/CGFloat(self.tabBarButtons.count)
        let buttonH = self.frame.size.height
        let buttonY = 0
        
        for index in 0...self.tabBarButtons.count - 1 {
//            取出按钮
            let button: UIButton = self.tabBarButtons[index] as! UIButton
//            2.设置按钮的frame
            let buttonX = CGFloat(index) * buttonW
            button.frame = CGRect(x: buttonX, y: CGFloat(buttonY), width: buttonW, height: buttonH)
             self.addSubview(button)
//            3.绑定tag
            button.tag = index
        }
        
    }
    
}
